enum.rb

```diff
@@ -1,19 +1,128 @@
+# frozen_string_literal: false
+
 module Enumerable
+  def my_each
+    return to_enum unless block_given?
 
-    def my_each
-        return to_enum unless block_given?
+    x = 0
+    while x < length
+      yield(to_a[x])
+      x += 1
+    end
+    self
+  end
 
-        x = 0
+  def my_each_with_index
+    return to_enum unless block_given?
 
-        while x < self.size
-            yield(to_a[x])
+    x = 0
+    while x < length
+      yield(self[x], x)
+      x += 1
+    end
+    self
+  end
 
-            x += 1
-        end
+  def my_select
+    return to_enum(:my_each) unless block_given?
 
-        self
+    arr = []
+    x = 0
+    while x < length
+      arr << self[x] if yield(self[x])
+      x += 1
+    end
+    arr
+  end
 
+  def my_all?(pat = nil)
+    result = true
+    if block_given?
+      my_each { |ele| result &= (yield ele) }
+    elsif pat
+      my_each { |ele| result &= pat == ele }
+    else
+      my_each { |ele| result &= ele }
     end
-end
```

Comment:
- [ ] should be identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when a class is passed as an argument returns true if all of the collection is a member of such class
```ruby
  array = [1,2,5,4,7]
  p array.my_all?(Integer) == array.all?(Integer) # false
```
- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex
```ruby
  array = %w[dog door rod blade]
  p array.my_all?(/c/) == array.all?(/c/) # false
```